pkg/agent/nodeportlocal/k8s/controller.go

```diff
@@ -0,0 +1,266 @@
+			return
+		}
+	}
+	podKey := podKeyFunc(pod)
+	c.queue.Add(podKey)
+
+}
+
+func (c *Controller) EnqueueObjDel(obj interface{}) {
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		var err error
+		pod, err = c.checkDeletedPod(obj)
+		if err != nil {
+			klog.Errorf("Got error while processing event update: %v", err)
+			return
+		}
+	}
+	podKey := podKeyFunc(pod)
+	c.queue.Add(podKey)
+
+}
+
+func (c *Controller) Worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	if key, ok := obj.(string); !ok {
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncPod(key); err == nil {
+		klog.V(2).Infof("Successfully processed key: %s, in queue", key)
+		c.queue.Forget(key)
+	} else {
+		c.queue.AddRateLimited(key)
+		klog.Errorf("Error syncing Pod %s, requeuing. Error: %v", key, err)
+	}
+	return true
+}
+
+func (c *Controller) addRuleForPod(pod *corev1.Pod) error {
+	podIP, nodeIP := pod.Status.PodIP, pod.Status.HostIP
+	if podIP == "" || nodeIP == "" {
+		return nil
+	}
+	podContainers := pod.Spec.Containers
+
+	for _, container := range podContainers {
+		for _, cport := range container.Ports {
+			port := int(cport.ContainerPort)
+			if c.portTable.RuleExists(podIP, port) {
+				continue
+			}
+			nodePort, err := c.portTable.AddRule(podIP, port)
+			if err != nil {
+				return err
+			}
+			assignPodAnnotation(pod, nodeIP, port, nodePort)
+		}
+	}
+	return nil
+}
+
+// HandleDeletePod removes rules from port table and
+// rules programmed in the system based on implementation type (e.g. IPTABLES)
+func (c *Controller) HandleDeletePod(key string) error {
+	klog.Infof("Got delete event for Pod: %s", key)
+	podIP, found := c.PodToIP[key]
+	if !found {
+		klog.Infof("IP address not found for Pod: %s", key)
+		return nil
+	}
+	data := c.portTable.GetDataForPodIP(podIP)
+	for _, d := range data {
+		err := c.portTable.DeleteRule(d.PodIP, int(d.PodPort))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// HandleAddUpdatePod handles Pod Add, Update events and updates annotation if required
+func (c *Controller) HandleAddUpdatePod(key string, obj interface{}) error {
+	newPod := obj.(*corev1.Pod).DeepCopy()
+	klog.Infof("Got add/update event for pod: %s", key)
+
+	podIP := newPod.Status.PodIP
+	if podIP == "" {
+		klog.Infof("IP address not found for pod: %s/%s", newPod.Namespace, newPod.Name)
+		return nil
+	}
+	c.PodToIP[key] = podIP
+
+	var err error
+	var updatePodAnnotation bool
+	newPodPorts := make(map[int]struct{})
+	newPodContainers := newPod.Spec.Containers
+	for _, container := range newPodContainers {
+		for _, cport := range container.Ports {
+			port := int(cport.ContainerPort)
+			newPodPorts[port] = struct{}{}
+			if !c.portTable.RuleExists(podIP, int(cport.ContainerPort)) {
+				err = c.addRuleForPod(newPod)
```

Comment:
Looks like there is some redundant code and computation between the nested loop here and the one in `addRuleForPod`.
maybe just move L183-L187 here, or remove L228-L232?